Show chapters in the bookmark dialog

ace20022 opened the pull request:
This is a replacement of  #6006
@notspiff @FernetMenta could you please have a look when you have some time. I've not reviewed it in detail till now, but have fixed some crashes, at least for win32.

notspiff committed:
added: ability to grab all chapter name and chapter positions from players

notspiff committed:
added: ability to specify position where thumb is taken from in DVDFileInfo

notspiff committed:
added: ability to specify where to extract thumb from in thumb extractor job

ace20022 committed:
added: show chapters in bookmark dialog (with thumbs)

Original version by Arne Morten Kvarving (notspiff).

notspiff committed:
added: if chapter is nameless, use a generic 'Chapter %u' string

notspiff committed:
added: support grabbing chapter name and position from bluray/dvds

notspiff commented:
looks sane. weird that i never saw the missing lock issue.

FernetMenta commented on xbmc/cores/dvdplayer/DVDPlayer.cpp:
```diff
@@ -4082,10 +4082,16 @@ int CDVDPlayer::GetChapter()
+  else
+  {
+    if (m_pDemuxer)
+      m_pDemuxer->GetChapterName(strChapterName, chapterIdx);
```
demuxer is not supposed to be accessed externally

notspiff replied:
dvdplayer cannot access its own member ?

FernetMenta replied:
sure but from dvdplayer's thread. external objects should not call into demuxer directly

notspiff replied:
so a vector in the state is needed then.

notspiff replied:
https://github.com/notspiff/kodi-cmake/commit/b964893b74c26219e0b35c4436528307fb3eb0b4 not tested.

ace20022 committed:
[dvdplayer] Fix wrong chapter numbers in dvd menus.

ace20022 commented:
@notspiff thanks much! I've added your commit and two changes that are necessary imo. The second one might prevent the access violations fixed in the first commit in menus but I've have not check that.

xhaggi commented on xbmc/cores/IPlayer.h:
```diff
@@ -179,7 +179,8 @@ class IPlayer
   virtual int  GetChapterCount()                               { return 0; }
   virtual int  GetChapter()                                    { return -1; }
-  virtual void GetChapterName(std::string& strChapterName)     { return; }
+  virtual void GetChapterName(std::string& strChapterName, int=-1) { return; }
```
should we add a name for the second paramter?

ace20022 committed:
[demuxer] set correct startpts when seeking chapters of discs.

ace20022 commented:
@notspiff @FernetMenta added some more commits, if you are okay with them I squash them keeping the original author.

notspiff commented:
I still dont get that lock. All actions are processed on the render/main thread - what exactly are you protecting with the lock, and from which thread?

ace20022 commented:
> I still dont get that lock. All actions are processed on the render/main thread - what exactly are you protecting with the lock, and from which thread?

stack trace without the lock: https://gist.github.com/ace20022/1f35b0ec0ec976db83d5
The access violation happens here:
```c++
  case CONTAINER_NUM_ITEMS:
    {
      unsigned int numItems = GetNumItems();
      if (numItems && m_items[0]->IsFileItem() && (boost::static_pointer_cast<CFileItem>(m_items[0]))->IsParentFolder())
        label = StringUtils::Format("%u", numItems-1);
      else
        label = StringUtils::Format("%u", numItems);
    }
    break;
```

m_items[0] is garbage.
log: https://gist.github.com/ace20022/f4abc2531e38523aa61e

I think I found a solution, I hope I can push it tonight with some other changes.

FernetMenta commented:
@ace20022 did you see my other message? OnRefreshList can only be called by main thread.

ace20022 commented:
Yes, I used g_windowManager.sendthreadmessage, iirc.

FernetMenta commented:
you should use GUIMessage. see example here: https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/windows/GUIWindowPVRBase.cpp#L80

ace20022 committed:
[VideoThumbLoader] Add parameter to skip querying/storing stream details when extracting thumbs.
This is useful for chapter thumbs extraction for example.

ace20022 commented:
@notspiff @FernetMenta  cleaned up bit. To sum up: without a lock it crashed, with the gfx lock it stutters and with https://github.com/ace20022/xbmc/commit/f5cadf26d737e32cd6b0e7781182ffa65127cb89 the ui freezes till the thumbs are loaded.
Test file with 32 chapters, rename to *.mkv:
![32ch](https://cloud.githubusercontent.com/assets/2150851/6205229/9f897e8c-b567-11e4-8e23-69a3ad0cafb2.jpg)

The last two commits try to lower the load, one can heat a whole house opening the dialog of a file with 800 chapters ;)

@FernetMenta are you okay with https://github.com/ace20022/xbmc/commit/691d6ab9a091777c99b51fd4a9046706f32e1be8 and my changes https://github.com/ace20022/xbmc/commit/42501bf128034bacd29e0c0da588c049aff763a3?

FernetMenta commented:
> are you okay with ace20022@691d6ab and my changes ace20022@42501bf? 

ok

how long does it freeze? what does block the render thread?

da-anda commented on language/English/strings.po:
```diff
@@ -12792,7 +12792,12 @@ msgctxt "#25009"
 
-#empty strings from id 25010 to 29799
+#: xbmc/video/dialogs/GUIDialogVideoBookmarks.cpp
+msgctxt "#25010"
```
could you evaluate where 21396 is used and if it's still needed or could be changed to this format? Also a contextual note would be nice. Something like this maybe:
#. Used to build chapter names for media like videos, dvd, bluray, ...

ace20022 replied:
search in files didn't find a usage.

da-anda replied:
ah, Confluence is using it - it's showing "chapter 2/10" in the dialogseekbar and VideoFullScreen. But I guess we could as well simply change 21396 to use %s and update Confluence accordingly. I'm just not sure if other skins also rely on the label which could become an issue with backwards compatibility. Are I** skins still compatible with Helix? @ronie ?

ronie replied:
string 21396 is used by many skins, so not a good idea to change that one imo.

> Are I** skins still compatible with Helix?
not 100% no. it's not that things are broken completely, far from that.
but there are some minor changes, mostly in the pvr area.

ace20022 commented:
> how long does it freeze?

Linear to the number of chapter thumbs extractions.

> what does block the render thread?

No idea, it seems that the list is is not updated incrementally.

FernetMenta commented:
can't you pause the debugger and see where it hangs?

FernetMenta commented on xbmc/video/dialogs/GUIDialogVideoBookmarks.cpp:
```diff
@@ -235,6 +239,30 @@ void CGUIDialogVideoBookmarks::OnRefreshList()
+    CFileItemPtr item(new CFileItem(name));
+    time = StringUtils::Format("chapter://%s/%i", path.c_str(), i);
+    std::string cachefile = CTextureCache::Get().GetCachedPath(CTextureCache::Get().GetCacheFile(time)+".jpg");
+    if (XFILE::CFile::Exists(cachefile))
```
hmm, I think this kicks off a job for a single item more than once. The job may be running already but the file is not cached

ace20022 commented on xbmc/video/dialogs/GUIDialogVideoBookmarks.cpp:
```diff
@@ -235,6 +239,30 @@ void CGUIDialogVideoBookmarks::OnRefreshList()
+    std::string cachefile = CTextureCache::Get().GetCachedPath(CTextureCache::Get().GetCacheFile(time)+".jpg");
+    if (XFILE::CFile::Exists(cachefile))
+      item->SetArt("thumb", cachefile);
+    else if (createThumbJob)
```
>hmm, I think this kicks off a job for a single item more than once. The job may be running already but the file is not cached

That's why I added ```createThumbJob```. With that it's quite unlikely and only when the dialog is closed and reopened while extracting.

FernetMenta replied:
assume yo have 10 chapters. at the first call to RefreshList is schedules 10 jobs. When the fist job finishes, it calls RefereshList again. At this time it will schedule another 9 jobs.

akva2 replied:
it won't. it will try, but those jobs won't be queued since they are already in there.

ace20022 replied:
No we do not because ```OnRefreshList(false)``` is called

FernetMenta replied:
ahh ok.

ace20022 commented:
> can't you pause the debugger and see where it hangs?

Is this a real or a rhetorical question?

Either way, it's waiting for 
https://github.com/xbmc/xbmc/blob/master/xbmc/Application.cpp#L4413 to return. More precise it's processing messages in the loop at https://github.com/xbmc/xbmc/blob/master/xbmc/ApplicationMessenger.cpp#L206 and each of that messages (i.e. chapter thumb) calls refresh list.

FernetMenta commented:
but it should not take very long to process OnRefreshList, shoulnd't it?

ace20022 commented:
Debug Print: CGUIDialogVideoBookmarks::OnRefreshList - measured 3098 ms to process 37 chapters
Debug Print: CGUIDialogVideoBookmarks::OnRefreshList - measured 2680 ms to process 37 chapters
Debug Print: CGUIDialogVideoBookmarks::OnRefreshList - measured 6995 ms to process 82 chapters
Debug Print: CGUIDialogVideoBookmarks::OnRefreshList - measured 5979 ms to process 82 chapters
Debug Print: CGUIDialogVideoBookmarks::OnRefreshList - measured 367924 ms to process 889 chapters

Each approx. between 30 and 250 ms.
There's a high chance that m_vecMessages.size() is increased inside the loop: https://github.com/xbmc/xbmc/blob/master/xbmc/ApplicationMessenger.cpp#L206 . In the 889 chapter case there were approx. 790 messages queued at a time for example.

FernetMenta commented:
hmm, maybe the database. open, read, close video db on every call is not a good idea. I would simplify it again by not trying to do two different paths in a this method.
1)
update the items in one method, cache items, minimize calls to db
2)
do not have render thread open db, have it update the ui when something had changed.

ace20022 commented:
Now there's only a short judder on open, create, delete bookmark. Also the wrong item gets updated if a bookmark is created/deleted while the chapter list is updated. Not sure how to fix this efficiently.

jenkins build this please

ace20022 commented:
jenkins build this please

ace20022 commented:
> Now there's only a short judder on open, create, delete bookmark.
That only happens with a large amount of chapters. I have no motivation left to fix that with this pr. 

> Also the wrong item gets updated if a bookmark is created/deleted while the chapter list is updated.
Fixed with last commit.

If both of you are happy I'll squash all related commits with
https://github.com/ace20022/xbmc/commit/ce9034c34fb98b017c0344a00ecbca1298b6daab

FernetMenta commented:
what happens when you open the dialog on a PI?

ace20022 commented:
This is a joke, isn't it?

FernetMenta commented:
I am sorry. what is so funny about this question? Maybe I miss something?

ace20022 commented:
> what happens when you open the dialog on a PI?

I'm almost sure that you will see a list of bookmarks items followed by chapter items.
If the pi is the new decision criterion for new features I'm out of here.

FernetMenta commented:
What do you mean by decision criteria? I was trying to say that it this makes the dialog unusable on low power platforms, we should make it configurable.

MartijnKaijser commented:
let's ask @popcornmix and @MilhouseVH how it runs

ace20022 commented:
> I was trying to say...

Then say so, I don't want to waste my precious free time on guessing games.

notspiff commented:
It is gonna be painful on toys while thumbs are extracted. This was always the case and is why regular extraction jobs are paused during playback. That would kinda suck here but a pi likely should suspend even these jobs to after playback to avoid stutter.

popcornmix commented:
I had a quick test with Pi1.
With the linked 32 chapter mkv file it wasn't too bad. Opening bookmark menu took about 20 seconds to populate all the thumbnails, but that file is only SD and virtually zero bitrate (17 Kbit/s).

I then tried a raw Blu-Ray mkv remux. Only 10 chapters. Not so good. Video/audio stuttered for tens of seconds before the thumbnails completed. It did work somewhat better than I'd expected.

Realistically on any ARM platform, you are not going to get a smooth experience trying to software decode 1080p video at the same time as playing the video with hardware decode. I'm also concerned about out-of-memory errors doing this (especially on a 256M Pi).

I'd suggest making this feature configurable in settings. Either disabled by default, or enabled but with overrides for specific platforms (e.g. in system/settings/rbp.xml).

You could possibly use the existing myvideos.extractflags / myvideos.extractthumb settings which are disabled by default on most Pi builds, but perhaps this should be a different setting.

da-anda commented:
can we extract these on import to library for new videos/installations?

NedScott commented:
I tested this on a little Android stick. If the only issue is the initial extraction, then I don't think that it is a problem if it is slow on a Pi. Pi users already know that initial scans and such take time. As long as there is a way to turn this off, or do as da-anda suggests (or both), then I think everyone would be happy.

Montellese commented:
IMO we should add an option for users to enable stream parsing and extraction etc during initial scanning. That way people with a fast machine can turn it on and people with a slow device can leave it off.

popcornmix commented:
Sure. No objections to the feature, or any problems with it being slow - it just needs a switch so users can enable/disable. Once there is a switch, it's easy to make that default to false on the Pi.

I think if you want to extract chapter thumbnails on library scan, then we should also support extracting the normal thumbnails on library scan. That has been rejected in the past.
It will make scanning an order of magnitude slower and people who just want to scan library so they can watch a newly added video will be annoyed by the extra time taken. Again, a switch will allow users who want it or don't want it to be happy.

ace20022 committed:
[VideoThumbLoader] Make == operator more strict.

ace20022 commented:
@popcornmix Thanks  a lot for your test. I've changed the process to sequential chapter thumb extraction. Could you please test the performance? If it is acceptable I would add 3 settings: no thumbs, serial extraction, fast as possible. Limiting the number of jobs might be reasonable for any platform, for my system, e.g., the cpu usage was 25% for 1 job at a time vs. 56% for 10 jobs at a time.

Extraction while importing can be done in another pr.

da-anda commented:
would serial extraction as fixed default be/feel too slow for fast machines? I'd not see an issue if the thumbs in the window are nicely fading in one by one as long as the UI is still responding fast etc.

Also, if opening the dialog is literally killing video playback on a PI, why don't we pause video playback when having the dialog opened? I mean, why would video have to continue? When you open the bookmark/chapter dialog you most likely want to jump to a specific bookmark or chapter - so you don't give a fuck about current playback anyway. And when playback is paused we could maybe even use the HW decoders to extract the thumbs?
That's btw the same with opening the audio or video OSD where in both cases you most likely change stuff and then jump back to where you have been - so we could as well just pause when opening a dialog in fullscreen mode. Netflix f.e. does also pause when you open up the audio/subtitle dialog and I never had any issue with this - I actually thought that this is a really good idea.

popcornmix commented:
Don't mind the idea of pausing automatically, but it won't help with using HW decoder to extract thumbs unless you are willing to close the original HW decoder and seek back to the last keyframe when you resume (as you will have lost the reference frames).

I think sometimes pausing would actually be useful, but not always - e.g. adjusting audio offset needs playback to be continuing, and changing audio track is nice to have feedback on what has changed.

da-anda commented:
@popcornmix well, the "hickup" on resume would only be noticeable IF there is a missing thumb in the bookmark dialog and it's closed again without going anywhere. In all other cases releasing the decoder shouldn't be an issue, right? Also resuming on a keyframe something like 3-5 secs earlier wouldn't be an issue for me personally - no idea what others think about it.

Regarding A/V osd dialogs - yes, sometimes it would make sense and sometimes not, true. But let's not get into these dialogs now and find a nice solution for the chapters.

ace20022 commented:
To get that in shape, where should the new setting be located? And how about a numeric setting, e.g., 0,1,2,...,30 parallel jobs.

FernetMenta commented:
I don't have any preferences on the setting. @da-anda @popcornmix what about you?

popcornmix commented:
It feels like a similar setting to "extract thumbnails and video information" which is in video/file lists.
Either that or add a new video/bookmarks section (which will be rather empty).

da-anda commented:
please no setting to configure the parallel extraction jobs. IMO simply stick to 1 job and create the thumbs one after another. On a fast machine this shouldn't cause too much delay and on low power ones it's also the best option we have and would allow us to use HW decoders. Ideally we show a placeholder for the time it takes to generate the thumbs and replace the placeholders one after another.

As for the settings location - yes, it's sort of related to the thumbs extraction in file lists, but it doesn't fit to the category. Creating a new category just for bookmarks also seems a bit too much for now, but @xhaggi and I have some ideas to make the big-jump forward/back seek step sizes configurable via GUI as well, and this by injecting fake chapters/bookmarks for which we need at least one additional chapter/bookmark related setting. So a "Chapters / Bookmarks" settings category would be fine from my end I guess (not having a better idea atm).

jjd-uk commented:
If you create a new category the then additive seeking "Skip steps" and "Skip delay" could be moved there after this goes in, my suggestion at the momemt would be to call this category "Navigation" but will have more of a think on possible names.

I'm ok with what da-anda suggests, have for now the:

Extract chapters and thumbs

with a one job at a time serial queue.

What maybe worthy of consideration in the future after this goes in is an additional setting (not sure in naming):

Extract thumb performance: low (default one job at a time)  / medium / high (whatever deemed a sensible number of parallel jobs)

But let's get this in so we can get user feedback on how it runs on various systems.

ace20022 committed:
[VideoBookmarks] Add a setting for chapter thumbnail extraction.

ace20022 committed:
[Settings/rpi] Disable chapter thumbnail extraction by default for the rpi.

ace20022 commented:
@da-anda @jjd-uk there's plenty of space in the file lists category, that's why I've added it there. Feel free to move it.
@popcornmix Have I added the override correctly?

Last chance to review or propose improvements.

jenkins build this please

da-anda commented:
@ace20022 I know that there is plenty of space in this category, but bookmarks unfortunately have nothing to do with file lists.

ace20022 commented:
Then please post a patch, I'll add it. I'm more interested in core changes and my list is long ;)

anaconda commented:
I've tested this on a Pi 2. Same results as popcornmix - apart from performance I've noticed that for 3D videos the thumbnail has 2 images (both sides).

da-anda commented:
@anaconda yes, thumbnail extraction for 3D movies is borked - always been like this (doesn't mean it couldn't get fixed)

ace20022 commented:
@anaconda 

> I've tested this on a Pi 2. Same results as popcornmix

So even with only 1 thread it kills video playback? And is the setting disabled by default?

popcornmix commented:
Yes, setting is disabled by default on pi, and thumbnails aren't created.
Enabling chapter thumbnails does work as expected.

NedScott commented:
Awesome!

I have some questions about how these new thumbnails work, so I can add this to the wiki (and generally just curious about what is currently possible or maybe planned for the future):

1. When a user exports their library as separate files, will that include the extracted thumbs?

2. Is it possible to manually change the thumbs, other than editing the generated cached image?

3. Is it possible to have Kodi extract chapter thumbnails before opening each movie individually? Maybe some kind of command that could be triggered over JSON-RPC or other method?

4. Will chapter thumbnails work over UPnP, like other Kodi artwork?

5. Will chapter thumbnails be accessible by remote apps/ remote web UIs?

ace20022 commented:
I'm afraid 5 times no; all would be nice to have. Next thing I'll look at are chapter files (xml, txt). But I first have to iterate ext. audio and bdj ;)

NedScott commented:
No worries, I figured this was just the start of things. This is good to know for people who ask. Thanks again!

MilhouseVH commented:
There's a problem streaming VIDEO_TS over smb:// with this PR - see [forum thread](http://forum.kodi.tv/showthread.php?tid=211501&pid=1945170#pid1945170) for details - basically it appears that VIDEO_TS streams over smb:// with multiple chapters are taking a huge performance hit as files are repeatedly opened and closed during playback - you don't have to do anything with bookmarks to see this problem, but it is caused by this PR as reverting this PR makes the problem go away.

I'm testing current Kodi master 492f9ba37624fa43518151786a5e89de863fa2b5 with a Pi2/OpenELEC.

My test VIDEO_TS is an "Extras" disc that has several "scenes". Most of the scenes consist of a single chapter and don't buffer or show the repeated open/close behaviour, however the one scene that has 6 chapters buffers significantly and has the repeated open/close behaviour. It may be a coincidence, but the fd appears to match the number of chapters.

"Extract chapter thumbnails" is disabled.

There is no problem with a DVD ISO, or MKV - the problem only seems to be with VIDEO_TS, and only over smb:// (not nfs://).

ace20022 commented:
I guess it's https://github.com/ace20022/xbmc/commit/572f64b4fb2918e63c76800227eef713ab4ea8e6#diff-14ed9651a2eb656ff75aea52d0912a00R4294

Not sure how to fix ad-hoc

ace20022 commented:
Will propose a fix later today.

Voyager1 commented:
@ace20022 this seems to break DVD menus completely: either stills are 'black', overlays are not shown. Any idea?

Voyager1 commented:
@ace20022 - disregard my comment. It's ffmpeg 2.6 which I was also testing in parallel. @FernetMenta I would advise against merging ffmpeg 2.6 until the DVD issues are resolved.

FernetMenta commented:
@Voyager1 I just tried a dvd iso and it's working fine here

ace20022 commented:
You scared me to death for a second ;)

Voyager1 commented:
sorry - it's not ffmpeg, it's not the chapter stuff - I found the cause and am working on a fix. Last good build was the one right before the Tuxbox removal, and I confirmed that https://github.com/xbmc/xbmc/commit/9aeb4a752359ff292112065fcf85bbaf4f8d4dc0 is the culprit. I suspect that some code in there was necessary for the good operation of DVD playback, although it's not obvious what it might be... I'll keep you posted.

popcornmix commented:
Have you got https://github.com/xbmc/xbmc/pull/6648?
That fixed all file operations breaking.

Voyager1 commented:
I arrived at the same conclusion just now - thanks that was it indeed :-)
